### ISSUE TITLE
pytorch_inference process crashes on MacOS

### ISSUE THREAD

**Issue by ajoliveira** [elastic/ml-cpp#2422]
ES Version: 8.5.1 (tar.gz)
MacOS: Monterey 12.6.1 (64GB, 8-core Intel Core i9)

When attempting to run example NLP model in our [docs](https://www.elastic.co/guide/en/machine-learning/8.5/ml-nlp-ner-example.html) the `pytorch_inference` process crashes when attempting to test with the sample text of `Elastic is headquartered in Mountain View, California.`.

Crash occurs regardless of:
- using Kibana UI and via Dev Tools
- debug/trace settings for `"logger.org.elasticsearch.xpack.ml"` or `"logger.org.elasticsearch.xpack.core.ml"`

Response is:
```
{
  "statusCode": 500,
  "error": "Internal Server Error",
  "message": "[status_exception: [status_exception] Reason: Error in inference process: [inference canceled as process is stopping]]: Error in inference process: [inference canceled as process is stopping]"
}
```

Example startup from `ps -ef` output:
>   501 13064 49123   0  9:08AM ttys000    0:02.94 ./pytorch_inference --validElasticLicenseKeyConfirmed=true --numThreadsPerAllocation=1 --numAllocations=1 --cacheMemorylimitBytes=265518993 --logPipe=/var/folders/m8/ss05dxq53y7_n_3wn08g0j_80000gn/T/elasticsearch-15093024931732039308/pytorch_inference_elastic__distilbert-base-uncased-finetuned-conll03-english_log_49096 --input=/var/folders/m8/ss05dxq53y7_n_3wn08g0j_80000gn/T/elasticsearch-15093024931732039308/pytorch_inference_elastic__distilbert-base-uncased-finetuned-conll03-english_input_49096 --inputIsPipe --output=/var/folders/m8/ss05dxq53y7_n_3wn08g0j_80000gn/T/elasticsearch-15093024931732039308/pytorch_inference_elastic__distilbert-base-uncased-finetuned-conll03-english_output_49096 --outputIsPipe --restore=/var/folders/m8/ss05dxq53y7_n_3wn08g0j_80000gn/T/elasticsearch-15093024931732039308/pytorch_inference_elastic__distilbert-base-uncased-finetuned-conll03-english_restore_49096 --restoreIsPipe --namedPipeConnectTimeout=10

Just in case, files in that directory are (never see any others created or if I should see any others):
```
➜  elasticsearch-15093024931732039308 l
total 32
-rw-r--r--  1 angelo  staff      0 Nov 16 19:16 jansi-2.4.0-556b068d611d35df-libjansi.jnilib.lck
-rwxr--r--  1 angelo  staff  15612 Nov 16 19:16 jansi-2.4.0-556b068d611d35df-libjansi.jnilib*
drwxr-xr-x  3 angelo  staff     96 Nov 16 19:16 geoip-databases/
```

Attaching log as well.

**Comment by ajoliveira** [comment]
ES log:
[home-cluster.log](https://github.com/elastic/ml-cpp/files/10032340/home-cluster.log)

**Comment by davidkyle** [comment]
I've reproduced this using the tar ball install for 8.5.1, 8.4.0 and 8.3.0. 

Inspecting the crash report in the macOS Console app the crash (SegV) occurred in `libtorch_cpu.dylib`

```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libtorch_cpu.dylib            	       0x11758cf36 0x112db3000 + 75341622
1   libtorch_cpu.dylib            	       0x11758445d 0x112db3000 + 75306077
2   libtorch_cpu.dylib            	       0x119069ba8 0x112db3000 + 103508904
3   libtorch_cpu.dylib            	       0x11757af76 0x112db3000 + 75267958
4   libtorch_cpu.dylib            	       0x1174c6eca 0x112db3000 + 74530506
5   libtorch_cpu.dylib            	       0x1173ae4d3 0x112db3000 + 73381075
6   libtorch_cpu.dylib            	       0x11739ffa2 0x112db3000 + 73322402
7   libtorch_cpu.dylib            	       0x112f8c9cc 0x112db3000 + 1939916
8   libtorch_cpu.dylib            	       0x112f7f678 0x112db3000 + 1885816
9   libtorch_cpu.dylib            	       0x112f7f33c at::native::structured_gelu_out_cpu::impl(at::Tensor const&, at::Tensor const&) + 124
10  libtorch_cpu.dylib            	       0x113b2786b 0x112db3000 + 14108779
11  libtorch_cpu.dylib            	       0x112dbb5f5 0x112db3000 + 34293
12  libtorch_cpu.dylib            	       0x1154f5322 0x112db3000 + 41165602
13  libtorch_cpu.dylib            	       0x1154ea578 0x112db3000 + 41121144
14  libtorch_cpu.dylib            	       0x1154dab77 0x112db3000 + 41057143
15  libtorch_cpu.dylib            	       0x115170266 torch::jit::Method::operator()(std::__1::vector<c10::IValue, std::__1::allocator<c10::IValue> >, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, c10::IValue, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, c10::IValue> > > const&) const + 534
16  pytorch_inference             	       0x109d02bb2 0x109cf9000 + 39858
17  pytorch_inference             	       0x109d020e3 0x109cf9000 + 37091
18  pytorch_inference             	       0x109d0aa3f 0x109cf9000 + 72255
19  pytorch_inference             	       0x109d1c91f 0x109cf9000 + 145695
20  pytorch_inference             	       0x109d0a2b7 0x109cf9000 + 70327
21  libMlCore.dylib               	       0x10a514633 ml::core::CStaticThreadPool::CWrappedTask::operator()() + 35
22  libMlCore.dylib               	       0x10a513ce7 ml::core::CStaticThreadPool::drainQueuesWithoutBlocking() + 199
23  libMlCore.dylib               	       0x10a513517 ml::core::CStaticThreadPool::shutdown() + 39
24  libMlCore.dylib               	       0x10a513792 ml::core::CStaticThreadPool::~CStaticThreadPool() + 18
25  libMlCore.dylib               	       0x10a54f8ec 0x10a42c000 + 1194220
26  pytorch_inference             	       0x109d059a7 0x109cf9000 + 51623
27  dyld                          	       0x10a8b052e start + 462
```

Looks like a null pointer dereference:

```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY
```

The crash does not occur with version 8.2, so it is likely to be related to the [PyTorch upgrade](https://github.com/elastic/ml-cpp/pulls?q=is%3Apr+is%3Aclosed+label%3Av8.3.0+label%3A%223rd+party+models%22) in 8.3. 

This may just affect macOS Monterey, it is unclear if it occurs on Big Sur or earlier.

**Comment by droberts195** [comment]
I tried this again with 8.5.3. As before, there's a SEGV when testing the "Elastic is headquartered in Mountain View, California." text with the `elastic/distilbert-base-uncased-finetuned-conll03-english` model.

The binaries we ship in releases are notarized.

The initial unnotarized binaries for each release are available. For 8.5.3 the relevant file is https://prelert-artifacts.s3.eu-west-1.amazonaws.com/maven/org/elasticsearch/ml/ml-cpp/8.5.3/ml-cpp-8.5.3-darwin-x86_64.zip.

I downloaded this file, deleted the `modules/x-pack-ml/platform/darwin-x86_64/controller.app` directory from the original Elasticsearch 8.5.3 install and replaced it by copying the `controller.app` directory out of the unzipped unnotarized download. I then unquarantined the new `controller.app` directory using `find controller.app | xargs xattr -d com.apple.quarantine`. Finally it's necessary to tell macOS not to worry about the lack of notarization by running `./controller.app/Contents/MacOS/pytorch_inference --help`, clicking "Cancel" when offered the opportunity to delete the suspect program, going to System Preferences -> Security and Privacy, and finally clicking on "Open Anyway" next to the message that `controller.app` was blocked.

After doing all this and restarting Elasticsearch the "Elastic is headquartered in Mountain View, California." example runs successfully. This is rather troubling, because the only difference in the binaries are that one has been notarized and one hasn't. So, something the notarization process does is causing the SEGV. 🙀

**Comment by droberts195** [comment]
There's a likely clue about what's happening in https://www.emaculation.com/forum/viewtopic.php?p=65599&sid=11553eba984a3038587550f8e5dcea68#p65599.

In order to notarize a program we have to sign it with `--options runtime` so that it uses the hardened runtime. By default that disallows JIT compilation. I suspect PyTorch might need the ability to do JIT compilation to be enabled.

If so, the the solution would be to sign using the additional argument `--entitlements entitlements.xml` where `entitlements.xml` contains this:

```
    <?xml version="1.0" encoding="UTF-8"?>
    <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
    <plist version="1.0">
        <dict>
            <key>com.apple.security.cs.allow-jit</key>
            <true/>
        </dict>
    </plist>
```

**Comment by droberts195** [comment]
Starting with the unnotarized binaries that work, I generated a self-certified development certificate using Xcode and signed them as follows:

```
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --strict "--verbose=9" ./controller.app/Contents/MacOS/*
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --strict "--verbose=9" ./controller.app/Contents/lib/*
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --strict "--verbose=9" ./controller.app
```

The problem then reproduces. This makes it pretty clear that PyTorch doesn't play well with the runtime hardening that `--options runtime` switches on.

Next, I tried signing everything again with the extra JIT entitlement:

```
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --entitlements entitlements.xml --strict "--verbose=9" ./controller.app/Contents/MacOS/*
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --entitlements entitlements.xml --strict "--verbose=9" ./controller.app/Contents/lib/*
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --entitlements entitlements.xml --strict "--verbose=9" ./controller.app
```

After that inference was possible once again.

Final test - just PyTorch given the extra entitlement:

```
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --strict "--verbose=9" ./controller.app/Contents/MacOS/*
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --strict "--verbose=9" ./controller.app/Contents/lib/*
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --entitlements entitlements.xml --strict "--verbose=9" ./controller.app/Contents/lib/*torch*
codesign -s "Apple Development: REDACTED" --deep --force --timestamp --options runtime --strict "--verbose=9" ./controller.app
```

The process crashes again.

So, it seems like the whole application needs to be given the JIT entitlement, not just the library that does it.

**Comment by droberts195** [comment]
This will be fixed in 8.6.0 and above by a change in the code signing process.

**Comment by droberts195** [comment]
Confirmed fixed in 8.6.0 BC10.
